File: algo/src/main/java/com/hartwig/actin/algo/doid/serialization/DatamodelCheckerFactory.java
package com.hartwig.actin.algo.doid.serialization;

import java.util.Map;

import com.google.common.collect.Maps;
import com.hartwig.actin.json.JsonDatamodelChecker;

import org.jetbrains.annotations.NotNull;

final class DatamodelCheckerFactory {

    private DatamodelCheckerFactory() {
    }

    @NotNull
    static JsonDatamodelChecker rootObjectChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("graphs", true);

        return new JsonDatamodelChecker("RootObject", map);
    }

    @NotNull
    static JsonDatamodelChecker graphsChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("nodes", true);
        map.put("edges", true);
        map.put("id", true);
        map.put("meta", true);
        map.put("equivalentNodesSets", true);
        map.put("logicalDefinitionAxioms", true);
        map.put("domainRangeAxioms", true);
        map.put("propertyChainAxioms", true);

        return new JsonDatamodelChecker("Graphs", map);
    }

    @NotNull
    static JsonDatamodelChecker nodeChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("type", false);
        map.put("lbl", false);
        map.put("id", true);
        map.put("meta", false);

        return new JsonDatamodelChecker("Node", map);
    }

    @NotNull
    static JsonDatamodelChecker edgeChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("sub", true);
        map.put("pred", true);
        map.put("obj", true);

        return new JsonDatamodelChecker("Edge", map);
    }

    @NotNull
    static JsonDatamodelChecker graphMetadataChecker() {
        Map<String, Boolean> map = Maps.newHashMap();

        map.put("xrefs", true);
        map.put("basicPropertyValues", true);
        map.put("version", false);
        map.put("subsets", true);

        return new JsonDatamodelChecker("GraphMetadata", map);
    }

    @NotNull
    static JsonDatamodelChecker logicalDefinitionAxiomChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("definedClassId", true);
        map.put("genusIds", true);
        map.put("restrictions", true);

        return new JsonDatamodelChecker("LogicalDefinitionAxiom", map);
    }

    @NotNull
    static JsonDatamodelChecker restrictionChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("propertyId", true);
        map.put("fillerId", true);

        return new JsonDatamodelChecker("Restriction", map);
    }

    @NotNull
    static JsonDatamodelChecker synonymChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("pred", true);
        map.put("val", true);
        map.put("xrefs", true);
        return new JsonDatamodelChecker("Synonym", map);
    }

    @NotNull
    static JsonDatamodelChecker definitionChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("xrefs", true);
        map.put("val", true);
        return new JsonDatamodelChecker("Definition", map);
    }

    @NotNull
    static JsonDatamodelChecker basicPropertyValueChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("pred", true);
        map.put("val", true);
        return new JsonDatamodelChecker("BasicPropertyValue", map);
    }

    @NotNull
    static JsonDatamodelChecker metadataXrefChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("val", true);

        return new JsonDatamodelChecker("MetadataXref", map);
    }

    @NotNull
    static JsonDatamodelChecker metadataChecker() {
        Map<String, Boolean> map = Maps.newHashMap();
        map.put("xrefs", false);
        map.put("synonyms", false);
        map.put("basicPropertyValues", false);
        map.put("definition", false);
        map.put("subsets", false);

        return new JsonDatamodelChecker("Metadata", map);
    }
}

File: common/src/main/kotlin/com/hartwig/actin/doid/serialization/DatamodelCheckerFactory.kt
package com.hartwig.actin.doid.serialization

import com.hartwig.actin.util.json.JsonDatamodelChecker

object DatamodelCheckerFactory {

    fun rootObjectChecker(): JsonDatamodelChecker {
        return allRequiredDatamodelChecker("RootObject", listOf("graphs"))
    }

    fun graphsChecker(): JsonDatamodelChecker {
        return JsonDatamodelChecker(
            "Graphs", mapOf("nodes" to true, "edges" to true, "id" to true, "meta" to true, "logicalDefinitionAxioms" to false)
        )
    }

    fun nodeChecker(): JsonDatamodelChecker {
        return JsonDatamodelChecker("Node", mapOf("type" to false, "lbl" to false, "id" to true, "meta" to false))
    }

    fun edgeChecker(): JsonDatamodelChecker {
        return allRequiredDatamodelChecker("Edge", listOf("sub", "pred", "obj"))
    }

    fun graphMetadataChecker(): JsonDatamodelChecker {
        return JsonDatamodelChecker(
            "GraphMetadata", mapOf("basicPropertyValues" to false, "version" to true)
        )
    }

    fun logicalDefinitionAxiomChecker(): JsonDatamodelChecker {
        return JsonDatamodelChecker(
            "LogicalDefinitionAxiom", mapOf("definedClassId" to true, "genusIds" to true, "restrictions" to false)
        )
    }

    fun restrictionChecker(): JsonDatamodelChecker {
        return allRequiredDatamodelChecker("Restriction", listOf("propertyId", "fillerId"))
    }

    fun synonymChecker(): JsonDatamodelChecker {
        return JsonDatamodelChecker(
            "Synonym", mapOf("pred" to true, "val" to true, "xrefs" to false, "synonymType" to false)
        )
    }

    fun definitionChecker(): JsonDatamodelChecker {
        return allRequiredDatamodelChecker("Definition", listOf("xrefs", "val"))
    }

    fun basicPropertyValueChecker(): JsonDatamodelChecker {
        return allRequiredDatamodelChecker("BasicPropertyValue", listOf("pred", "val"))
    }

    fun metadataXrefChecker(): JsonDatamodelChecker {
        return allRequiredDatamodelChecker("MetadataXref", listOf("val"))
    }

    fun metadataChecker(): JsonDatamodelChecker {
        return JsonDatamodelChecker(
            "Metadata",
            listOf("xrefs", "synonyms", "basicPropertyValues", "definition", "subsets", "deprecated", "comments").associateWith { false }
        )
    }

    private fun allRequiredDatamodelChecker(name: String, properties: List<String>) =
        JsonDatamodelChecker(name, properties.associateWith { true })
}
